3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -368,6 +360,68 @@ class column_base : public column_reader,
     return *stream_;
   }
 
+ protected:
+  column_header& mutable_header() { return hdr_; }
+  void reset_stream(const index_input* stream) { stream_ = stream; }
+
+  bool allocate_buffered_memory(size_t size) {
+    assert(size < static_cast<size_t>(std::numeric_limits<int64_t>::max()));
+    assert(buffered_memory_accounter_);
+    if (!buffered_memory_accounter_(static_cast<int64_t>(size))) {
+      // use string to have always null-terminated name
+      std::string col_name{"<anonymous>"};
+      if (!name().null()) {
+        col_name = name();
+      }
+      IR_FRMT_WARN(
+          "Failed to allocate memory for buffered column id %d name: %s of "
+          "size " IR_SIZE_T_SPECIFIER,
+          header().id, col_name.c_str(), size);
+      return false;
+    }
+    // should be only one alllocation
+    assert(column_data_.empty());
+    column_data_.resize(size);
+    return true;
+  }
+
+  size_t calculate_bitmap_size(size_t file_len,
+                               range<std::unique_ptr<column_reader>>
+                                   next_sorted_columns) const noexcept {
+    if (!header().docs_index) {
+      return 0;
+    }
+    for (const auto& c : next_sorted_columns) {
+      auto column = static_cast<column_base*>(c.get());
+      assert(column != nullptr);
+      if (column->header().docs_index) {
+        file_len = column->header().docs_index;
+        break;
+      }
+    }
+    assert(header().docs_index < file_len);
+    return file_len - header().docs_index;
+  }
+
+  void store_bitmap_index(size_t bitmap_size, size_t buffer_offset,
+                          remapped_bytes_ref_input::mapping* mapping,
+                          column_header& hdr, index_input& in) {
+    assert(bitmap_size);
+    assert(hdr.docs_index);
+    in.read_bytes(hdr.docs_index, column_data_.data() + buffer_offset,
+                  bitmap_size);
+    if (is_encrypted(hdr)) {
+      assert(mapping);
+      mapping->emplace_back(hdr.docs_index, buffer_offset);
```

Comment:
If the `emplace_back` throws here, is the object left in an inconsistent state and needs to be cleaned up?

Reply:
Yes. Column is in the undefined state. But still allocated memory would be freed in dtor. Anyway exception at that stage will render entire index segment as failed. So I assume here nothing should be done to restore column state.